### ISSUE TITLE
[FEATURE] Remove global state

### ISSUE THREAD

**Issue by Evrey** [Alexhuszagh/rust-lexical#45]
## Problem
Currently, some settings like the current expected exponent character are global state. This can be anything from inconvenience to tricky issue for projects parsing multiple languages.

## Solution
The `format` feature of `lexical-core` added integer-packed settings which you have to pass to all format parsing functions. I suggest doing a similar thing for everything else, but by passing in a struct reference. Integer-packing works for things like the exponent characters, but fails for e.g. `set_inf_string`. This is still C-API-friendly. C libs are just forced to put `strlen` next to their `*const c_char`s.

## Prerequisites

- lexical version : `0.7.*`
- lexical compilation features used: `format`, `radix`, `correct`

## Alternatives
Uhh… not doing any of this? Maybe packing said strings into something like `staticvec::StaticString` inlined into the struct? But I don't think that gives any benefit. Another idea would be to always enable `format` and include `format`'s bit-packed settings into that one struct.

**Comment by myrrlyn** [comment]
Hi. I'm the new maintainer until the original Alex is able to return.

I am not opposed to this. I will need to look through the crate and its public APIs in order to get familiar with how they are currently used. I think, from first impression, that the correct thing to do would be to remove the mutability and use the current state as a default, and provide sibling APIs that allow providing your own information rather than using the crate-internal defaults.

I probably won't be able to work on this much before August. I am not asking you for a PR, but if you would like to draft this before I am able to, I certainly would welcome the collaboration.

Thank you for your patience.

**Comment by Alexhuszagh** [comment]
@RazrFalcon I think as long as we use `&'static string` or similar, we should be fine. There's trivial overhead if we're using 3 pointers (for all 3 string representations) and minimizing the copying compared to the rest of the algorithm. Also, the `FormatNumber` API should be supported fully from the C end, so it's pretty trivial to ask for a `const char*` and a length for `NaN`, `inf`, and `infinity`.

This also should have already been addressed with the [ParseFloatOptionsBuilder](https://github.com/Alexhuszagh/rust-lexical/blob/v0.8/lexical-core/src/util/options.rs#L294), which allows you to ignore all the mutable, global state, which was being phased in for v0.8 before I had to step away for mental health reasons.

**Comment by Evrey** [comment]
@myrrlyn Sadly i don't have the spare time to work through this codebase and draft something, but i am patient. August? No problem. I knew i would be waiting for at least a year when i wrote my issues. =) I like this default solution you propose.

@Alexhuszagh Oh my, best wishes for your health situation.

**Comment by Alexhuszagh** [comment]
@RazrFalcon Does the proposed option of `ParseFloatOptionsBuilder` work for you. It also dramatically simplifies creating `FloatOptions`. A sample of the use is as follows. The global state will be removed, and temporarily be deprecated for use `ParseFloatOptionsBuilder`.

https://github.com/Alexhuszagh/rust-lexical/blob/v0.8/lexical-core/src/util/options.rs#L305-L396

```rust
use lexical_core::{ParseFloatOptions, Builder, Buildable};

pub fn main() {
    let options = ParseFloatOptions::builder()
        .lossy(true)
        .exponent_char(b'e')
        .nan_string(b"NaN")
        .inf_string(b"Inf")
        .infinity_string(b"Infinity")
        .build()
        .unwrap();
}
```

**Comment by Alexhuszagh** [comment]
@RazrFalcon I've added `Parse*Options` and `Write*Options` (new structs to accommodate removing global state), made all their member functions `const fn` (along with builders, which allow the safe, no-panic creation of these options), and then integrated the `atof` functions in terms of these. The API is the same as above, and I've spent a lot of time to ensure this has no discernible performance impact (a lot harder than I initially imagined).

**Comment by Alexhuszagh** [comment]
Implemented as of [lexical v6.0.0](https://crates.io/crates/lexical/6.0.0) and [lexical-core v0.8.0](https://crates.io/crates/lexical-core/0.8.0). No global state exists in any capacity anymore.
